runtime/etc/mli/mli_server_runtime.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mli_common_code.c"
+#include <time.h>
+
+//
+// The definition of this is generated by the compiler.
+// TODO: Make extern later?
+//
+int64_t chpl_mli_sdispatch(int64_t id);
+
+//
+// The definition of this is generated by the compiler.
+// TODO: Make extern later?
+//
+void chpl_mli_server_set_conf(void);
+
+struct chpl_mli_gen_conf {
+
+  int debug;
+
+};
+
+//
+// These can be set to TRUE by generated code via a generated setter.
+//
+struct chpl_mli_gen_conf chpl_server_conf = {
+  .debug  = 0,
+};
+
+//
+// A little bit of global context will do us good in the long run.
+//
+struct chpl_mli_context chpl_server;
+
+void chpl_mli_server_init(struct chpl_mli_context* server) {
+  if (server->context) { return; }
+
+  server->context = zmq_ctx_new();
+  server->main    = zmq_socket(server->context, ZMQ_REP);
+  server->arg     = zmq_socket(server->context, ZMQ_REP);
+  server->res     = zmq_socket(server->context, ZMQ_REQ);
+
+  return;
+}
+
+void chpl_mli_server_deinit(struct chpl_mli_context* server) {
+  if (!server->context) { return; }
+
+  // TODO: Set socket LINGER to 0 to prevent blocking.
+  zmq_ctx_destroy(server->context);
+  server->context = NULL;
+
+  return;
+}
+
+//
+// TODO: Write a more elegant wrapper around printf.
+//
+// Only valid in the context of the server main loop for now. Expects the
+// "chpl_server_conf" global to be present.
+//
+#define chpl_mli_sdebugf(fmt, ...)              \
+  do {                                          \
+    if (!chpl_server_conf.debug) { break; }     \
+      printf("%s ", "[Server]");                \
+      printf(fmt, __VA_ARGS__);                 \
+  } while (0)
```

Comment:
While I'm thinking about it, this seems really weird to me.  Like, it's a do-while loop that executes at most once?  Shouldn't that only be the if statement?

Reply:
It's a technique for writing macros that causes the compiler to complain if I don't terminate the macro invocation with a semicolon. If I used an if statement by itself, then the semicolon would be ignored as a "NOP statement", which might cause problems when nesting calls to macro functions.

Reply:
Ah, okay.  Thanks for the explanation!  I'm glad I asked